packages/flutter/lib/src/painting/image_stream.dart

```diff
@@ -2,35 +2,140 @@
+    _handles[image] ??= ImageHandle._(image);
+    return _handles[image]!;
+  }
+
+  static final Expando<ImageHandle> _handles = Expando<ImageHandle>();
+
+  final ui.Image _image;
+  int _refCount = 0;
+
+  /// The number of references to this [ImageHandle].
+  ///
+  /// This always returns null if asserts are disabled, and is only meant for
+  /// use in debug mode or tests.
+  int? get debugRefCount {
+    int? returnValue;
+    assert(() {
+      returnValue = _refCount;
+      return true;
+    }());
+    return returnValue;
+  }
+
+  bool _debugDisposed = false;
+
+  /// Whether all holders of this handle have called [dispose].
+  ///
+  /// This always returns false if asserts are disabled, and is only meant for
+  /// use in debug mode or tests.
+  bool get debugDisposed {
+    bool returnValue = false;
+    assert(() {
+      returnValue = _debugDisposed;
+      return true;
+    }());
+    return returnValue;
+  }
+
+  /// Called by [ImageInfo.obtainHandle].
+  void _ref() {
+    assert(!_debugDisposed, 'Object has been disposed.');
+    _refCount += 1;
+  }
+
+  /// Release this image handle.
+  ///
+  /// Once all outstanding image handles have been disposed, the underlying
+  /// [ui.Image] will be disposed as well. Further attempts to use this handle
+  /// will throw.
+  ///
+  /// Clients must call this method once and only once for each call to
+  /// [ImageInfo.obtainHandle].
+  void dispose() {
+    assert(!_debugDisposed, 'Object has been disposed');
+    _refCount -= 1;
+    if (_refCount == 0) {
+      assert(() {
+        _debugDisposed = true;
+        return true;
+      }());
+      _image.dispose();
+    }
+  }
+}
+
 /// A [dart:ui.Image] object with its corresponding scale.
 ///
 /// ImageInfo objects are used by [ImageStream] objects to represent the
 /// actual data of the image once it has been obtained.
+///
+/// Before accessing the [image], callers must use [obtainImageHandle] to get
+/// an [ImageHandle]. When the caller no longer needs the image, they should
+/// call [ImageHandle.dispose].
 @immutable
 class ImageInfo {
-  /// Creates an [ImageInfo] object for the given [image] and [scale].
+  /// Creates an [ImageInfo] object for the given [imageHandle] and [scale].
   ///
-  /// Both the image and the scale must not be null.
+  /// The [imageHandle] and [scale] parameters must not be null.
   ///
-  /// The tag may be used to identify the source of this image.
-  const ImageInfo({ required this.image, this.scale = 1.0, this.debugLabel })
-    : assert(image != null),
-      assert(scale != null);
+  /// A client of this class must call [obtainHandle] before using the [image],
+  /// and then call [ImageHandle.dispose] when they will no longer need the
+  /// [image]. Once all clients have disposed of their handles, the [image] will
+  /// be in a disposed state as well and will no longer be usable.
+  ///
+  /// The [debugLabel] may be used to identify the source of this image.
+  ImageInfo({
```

Comment:
FYI: removing const is a breaking change.

Reply:
Only if it breaks a contributed test or a test in google3.

Reply:
yeah, that generally seems unlikely in this case - its not really a commonly used type